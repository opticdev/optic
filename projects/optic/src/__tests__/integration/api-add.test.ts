import {
  test,
  expect,
  describe,
  jest,
  beforeEach,
  afterEach,
} from '@jest/globals';
import {
  runOptic,
  setupWorkspace,
  normalizeWorkspace,
  setupTestServer,
  run,
} from './integration';
jest.setTimeout(30000);

function sanitizeOutput(out: string) {
  return out.replace(/[a-zA-Z0-9]{8}:/g, 'COMMIT-HASH:');
}

setupTestServer(({ url, method }) => {
  if (method === 'POST' && /\/api\/specs\/prepare$/.test(url)) {
    return JSON.stringify({
      upload_id: '123',
      spec_url: `${process.env.BWTS_HOST_OVERRIDE}/special-s3-route/spec`,
      sourcemap_url: `${process.env.BWTS_HOST_OVERRIDE}/special-s3-route/sourcemap`,
    });
  } else if (method === 'POST' && /\/api\/specs$/.test(url)) {
    return JSON.stringify({
      id: 'spec-id',
    });
  } else if (method === 'GET' && /\/api\/token\/orgs/.test(url)) {
    return JSON.stringify({
      organizations: [{ id: 'org-id', name: 'org-blah' }],
    });
  } else if (method === 'GET' && /\/api\/ruleset-configs/.test(url)) {
    // a return value means it exists
    return JSON.stringify({});
  } else if (method === 'POST' && /\/api\/api/.test(url)) {
    return JSON.stringify({ id: 'api-id' });
  }

  return JSON.stringify({});
});

describe('optic api add', () => {
  let oldEnv: any;
  beforeEach(() => {
    oldEnv = { ...process.env };
    process.env.LOG_LEVEL = 'info';
    process.env.OPTIC_ENV = 'test';
  });

  afterEach(() => {
    process.env = { ...oldEnv };
  });

  describe.each([
    [
      'git',
      {
        repo: true,
        commit: true,
      },
    ],
    [
      'no vcs',
      {
        repo: true, // To force a no vcs - we need to init an empty git repo (otherwise it uses the optic git repo)
        commit: false,
      },
    ],
  ])('%s - no cli interaction', (vcs, setupOptions) => {
    test('discover one file with history depth', async () => {
      process.env.OPTIC_TOKEN = 'something';
      const workspace = await setupWorkspace('api-add/one-file', setupOptions);
      if (vcs === 'git') {
        // add another commit
        await run(
          `touch ./hello.yml && git add . && git commit -m 'another one'`,
          false,
          workspace
        );
      }
      const { combined, code } = await runOptic(
        workspace,
        'api add ./spec.yml --standard blah'
      );

      expect(code).toBe(0);
      expect(
        normalizeWorkspace(workspace, sanitizeOutput(combined))
      ).toMatchSnapshot();
      // expect spec to have added yml files
      const { code: specCode, combined: specCombined } = await run(
        'cat ./spec.yml',
        false,
        workspace
      );
      expect(specCode).toBe(0);
      expect(normalizeWorkspace(workspace, specCombined)).toMatchSnapshot();
    });

    test('discover all files in repo', async () => {
      process.env.OPTIC_TOKEN = 'something';

      const workspace = await setupWorkspace(
        'api-add/many-files',
        setupOptions
      );

      const { combined, code } = await runOptic(
        workspace,
        'api add --all --standard blah'
      );

      expect(code).toBe(0);
      expect(
        normalizeWorkspace(workspace, sanitizeOutput(combined))
      ).toMatchSnapshot();
    });
  });
});
