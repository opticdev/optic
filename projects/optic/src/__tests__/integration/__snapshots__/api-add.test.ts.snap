// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`optic api add git - no cli interaction discover all files in a folder 1`] = `
"

[1m[90mLooking for OpenAPI specs in directory $workspace$/nested[39m[22m
[1m[32mâœ”[39m[22m [1m[34ma spec[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m

[34m[1mBackfilling API history, you can exit at any time (\`ctrl + c\`) and finish this later.[22m[39m

[1m[90mBackfilling history for nested/speccopy2.yml[39m[22m
[32mâœ”[39m [1m[34ma spec[39m[22m history backfilled

[34m[1mx-optic-url has been added to newly tracked specs. You should commit these changes.[22m[39m

[34m[1mSetup CI checks by running "optic ci setup"[22m[39m
"
`;

exports[`optic api add git - no cli interaction discover all files in repo 1`] = `
"

[1m[90mLooking for OpenAPI specs in directory $workspace$[39m[22m
[1m[32mâœ”[39m[22m [1m[34mEmpty[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m
[1m[32mâœ”[39m[22m [1m[34ma spec[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m
[1m[32mâœ”[39m[22m [1m[34ma spec[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m

[34m[1mBackfilling API history, you can exit at any time (\`ctrl + c\`) and finish this later.[22m[39m

[1m[90mBackfilling history for example-api-v0.json[39m[22m
[32mâœ”[39m [1m[34mEmpty[39m[22m history backfilled

[1m[90mBackfilling history for nested/speccopy2.yml[39m[22m
[32mâœ”[39m [1m[34ma spec[39m[22m history backfilled

[1m[90mBackfilling history for spec.yml[39m[22m
[32mâœ”[39m [1m[34ma spec[39m[22m history backfilled

[34m[1mx-optic-url has been added to newly tracked specs. You should commit these changes.[22m[39m

[34m[1mSetup CI checks by running "optic ci setup"[22m[39m
"
`;

exports[`optic api add git - no cli interaction discover all files in repo using --all 1`] = `
"

[1m[90mLooking for OpenAPI specs in directory $workspace$[39m[22m
[1m[32mâœ”[39m[22m [1m[34mEmpty[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m
[1m[32mâœ”[39m[22m [1m[34ma spec[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m
[1m[32mâœ”[39m[22m [1m[34ma spec[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m

[34m[1mBackfilling API history, you can exit at any time (\`ctrl + c\`) and finish this later.[22m[39m

[1m[90mBackfilling history for example-api-v0.json[39m[22m
[32mâœ”[39m [1m[34mEmpty[39m[22m history backfilled

[1m[90mBackfilling history for nested/speccopy2.yml[39m[22m
[32mâœ”[39m [1m[34ma spec[39m[22m history backfilled

[1m[90mBackfilling history for spec.yml[39m[22m
[32mâœ”[39m [1m[34ma spec[39m[22m history backfilled

[34m[1mx-optic-url has been added to newly tracked specs. You should commit these changes.[22m[39m

[34m[1mSetup CI checks by running "optic ci setup"[22m[39m
"
`;

exports[`optic api add git - no cli interaction discover one file with history depth 1`] = `
"

[1m[90mAdding API $workspace$/spec.yml[39m[22m
[1m[32mâœ”[39m[22m [1m[34ma spec[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m

[34m[1mBackfilling API history, you can exit at any time (\`ctrl + c\`) and finish this later.[22m[39m

[1m[90mBackfilling history for spec.yml[39m[22m
[32mâœ”[39m [1m[34ma spec[39m[22m history backfilled

[34m[1mx-optic-url has been added to newly tracked specs. You should commit these changes.[22m[39m

[34m[1mSetup CI checks by running "optic ci setup"[22m[39m
"
`;

exports[`optic api add git - no cli interaction discover one file with history depth 2`] = `
"openapi: 3.0.3
x-optic-url: "https://app.useoptic.com/organizations/org-id/apis/api-id"
info:
  title: a spec
  description: The API
  version: 0.1.0
paths:
  /filler_route:
    post:
      operationId: create
      responses:
        "201":
          description: Created successfully
          content:
            application/json:
              schema:
                type: object
                properties:
                  id:
                    type: string
                    format: uuid
                    example: d5b640e5-d88c-4c17-9bf0-93597b7a1ce2
"
`;

exports[`optic api add no vcs - no cli interaction discover all files in a folder 1`] = `
"

[1m[90mLooking for OpenAPI specs in directory $workspace$/nested[39m[22m
[1m[32mâœ”[39m[22m [1m[34ma spec[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m

[34m[1mx-optic-url has been added to newly tracked specs. You should commit these changes.[22m[39m

[34m[1mSetup CI checks by running "optic ci setup"[22m[39m
"
`;

exports[`optic api add no vcs - no cli interaction discover all files in repo 1`] = `
"

[1m[90mLooking for OpenAPI specs in directory $workspace$[39m[22m
[1m[32mâœ”[39m[22m [1m[34mEmpty[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m
[1m[32mâœ”[39m[22m [1m[34ma spec[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m
[1m[32mâœ”[39m[22m [1m[34ma spec[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m

[34m[1mx-optic-url has been added to newly tracked specs. You should commit these changes.[22m[39m

[34m[1mSetup CI checks by running "optic ci setup"[22m[39m
"
`;

exports[`optic api add no vcs - no cli interaction discover all files in repo using --all 1`] = `
"

[1m[90mLooking for OpenAPI specs in directory $workspace$[39m[22m
[1m[32mâœ”[39m[22m [1m[34mEmpty[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m
[1m[32mâœ”[39m[22m [1m[34ma spec[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m
[1m[32mâœ”[39m[22m [1m[34ma spec[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m

[34m[1mx-optic-url has been added to newly tracked specs. You should commit these changes.[22m[39m

[34m[1mSetup CI checks by running "optic ci setup"[22m[39m
"
`;

exports[`optic api add no vcs - no cli interaction discover one file with history depth 1`] = `
"

[1m[90mAdding API $workspace$/spec.yml[39m[22m
[1m[32mâœ”[39m[22m [1m[34ma spec[39m[22m is now being tracked.
  [1mView: [4mhttps://app.useoptic.com/organizations/org-id/apis/api-id[24m[22m

[34m[1mx-optic-url has been added to newly tracked specs. You should commit these changes.[22m[39m

[34m[1mSetup CI checks by running "optic ci setup"[22m[39m
"
`;

exports[`optic api add no vcs - no cli interaction discover one file with history depth 2`] = `
"openapi: 3.0.3
x-optic-url: "https://app.useoptic.com/organizations/org-id/apis/api-id"
info:
  title: a spec
  description: The API
  version: 0.1.0
paths:
  /filler_route:
    post:
      operationId: create
      responses:
        "201":
          description: Created successfully
          content:
            application/json:
              schema:
                type: object
                properties:
                  id:
                    type: string
                    format: uuid
                    example: d5b640e5-d88c-4c17-9bf0-93597b7a1ce2
"
`;
